### DISCUSSION TITLE
TLS Fatal Alert Protocol Version in middle of stream

### DISCUSSION THREAD

**Discussion by mark99i** [openssl/openssl#23477] [Q&A]
I am developing an application to interact with a windows hyper v service, the stack of which is TCP -> TLS -> RDP and I am faced with the problem that the client (application) at some point sends the TLS Fatal Alert Protocol Version to the server

however, based on logic, this error can occur only when the connection is negotiated, but it occurs to me after the agreement and after a certain amount of useful data

the aggravating factor is that I use the high-level python language (I don't have enough experience to debug low-level openssl code) and this error only occurs on linux (initially, I developed on windows and everything is ok there), I checked many versions starting with the old 1.0.x (debian9), ending with the master branch

to build the master branch, I used the following dockerfile: https://gist.github.com/mark99i/9a75a45300ebe6de7a663ab445b453a1

from the application side, it looks as if the connection just crashes without errors, however, I tried to dump the keys through the env SSLKEYLOGFILE and got this at the time of the break, here you can see that the error occurs after the payload of the high-level protocol
![photo_2023-06-13_13-32-50](https://github.com/openssl/openssl/assets/46539611/4b16ed36-49c0-4133-8a64-e33f968342e4)
I also tried different encryption options and protocol versions: TLS 1.2, TLS 1.3, with and without DH, AES-128, RC4 and their combinations until the server rejected the handshake. in all cases, when the handshake was completed successfully, the bug was reproduced

what could be the problem and how can I help find it? will a traffic dump help with the alert?

**Comment by mattcaswell** [comment]
It's unclear from your description: are you using OpenSSL as a server or as a client?  i.e. is it your application that sends the alert (via OpenSSL), or the remote peer? Is the peer also using OpenSSL or something else?

What version of TLS has been negotiated?

> however, based on logic, this error can occur only when the connection is negotiated, but it occurs to me after the agreement and after a certain amount of useful data

The error can occur at any time, e.g. if a record header for an application data record does not have the expected version number in it. This might occur (for example) if data has got corrupted somewhere. You might examine the last record sent from the server to the client immediately before the client sends the alert and check that it looks sane.

**Comment by mark99i** [comment]
> It's unclear from your description: are you using OpenSSL as a server or as a client? i.e. is it your application that sends the alert (via OpenSSL), or the remote peer? Is the peer also using OpenSSL or something else?

the client sends an alert to the server, I am working on the client (application). 
what the remote side uses is unknown, it's a windows service, maybe openssl, maybe an internal stack


by default, the session is negotiated on TLS 1.2: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (0xc030)
manually the server agrees TLS1.3 and 1.2, without DH and AES-128, as far as I understood it does not affect.

the only thing that affects is the change of the client's platform. the client's problem manifests itself only on linux, on windows with the same version openssl works without errors

> The error can occur at any time, e.g. if a record header for an application data record does not have the expected version number in it. This might occur (for example) if data has got corrupted somewhere. You might examine the last record sent from the server to the client immediately before the client sends the alert and check that it looks sane.

tcp, as far as I know, should guarantee integrity and sequence of bytes, it's hardly a simple packet corruption. even if that were the case, it wouldn't be replayed every time. on the contrary, I have never managed to have a session with active traffic live longer than a few seconds...

I looked at the field 'Version' and in all packages of the session, its value was 1.2, except for the first client hello - there is 1.0. I couldn't find anything strange.
![image](https://github.com/openssl/openssl/assets/46539611/41076699-feb8-44bc-bd1a-34b7e555fa30)

**Comment by mark99i** [comment]
@mattcaswell 
I ran the tests again, in each case, I did several sessions

The environment and the application code are the same, except for the platforms (windows native, linux docker)
Python 3.12 + OpenSSL 3.0.11 19 Sep 2023
automatic negotiation on TLS1.2
Cipher Suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (0xc030)

[linux-alert.pcap.zip](https://github.com/openssl/openssl/files/14168945/linux-alert.pcap.zip)
[windows-working.pcap.zip](https://github.com/openssl/openssl/files/14168947/windows-working.pcap.zip)

[keys.log](https://github.com/openssl/openssl/files/14168943/keys.log) contains session keys (SSLKEYLOGFILE)

why can there be such behavior?

**Comment by mark99i** [comment]
I compiled the openssl master from the source code and now received a TLS Alert Internal Error in one of the sessions
Environment: debian:sid + OpenSSL 3.3.0-dev + Python 3.12.1+ (branch 3.12)

![image](https://github.com/openssl/openssl/assets/46539611/29e2e3f3-59c3-4a04-8210-bc6703d14cdc)
[linux-openssl-3.3.0-dev.pcap.zip](https://github.com/openssl/openssl/files/14169262/linux-openssl-3.3.0-dev.pcap.zip)
[keys.log](https://github.com/openssl/openssl/files/14169263/keys.log)

**Comment by mark99i** [comment]
@mattcaswell Any ideas?

**Reply by mattcaswell** [reply]
The linux pcap file looks quite strange to me. Earlier on in the communication I see various lines which Wireshark reports as TLSv1.2, but with info "Ignored Unknown Record" - and indeed the TLS data is total trash. Reading such a trash record could certainly result in the protocol version alert being raised. Strangely though I also see these lines in the working windows capture that you sent. So its unclear to me why it works there.

**Reply by mark99i** [reply]
Unfortunately, I can't change the server side, in which case is there any way to understand why it works on windows? 
I think I can take a traffic dump from the intermediate router, maybe the dump from the destination machine is incomplete (Previous segment(s) not captured (common at capture start))

**Comment by mark99i** [comment]
@mattcaswell I took a dump from another node and there are no more `Ignore Unknown Record` entries in them, however the error is reproducible
also now there is another error: Unexpected Message (10)

[keys.log](https://github.com/openssl/openssl/files/14211033/keys.log)
[linux-master-alert-08.02.pcap.zip](https://github.com/openssl/openssl/files/14211034/linux-master-alert-08.02.pcap.zip)
[win-working-08.02.pcap.zip](https://github.com/openssl/openssl/files/14211035/win-working-08.02.pcap.zip)
![image](https://github.com/openssl/openssl/assets/46539611/9cf9a5f7-3f8f-490b-8924-f792c0b2ad06)

**Comment by mark99i** [comment]
@mattcaswell Any ideas?

**Comment by mark99i** [comment]
@t8m Hello, could you see what the problem might be?

**Reply by t8m** [reply]
Unfortunately I have no idea.

**Comment by mark99i** [comment]
@t8m @mattcaswell @paulidale 
hello, can you tell me what information I need to provide to search for what could be the problem?

**Comment by nhorman** [comment]
I assume that your client is sending the alert during/after a call to SSL_read or SSL_handle_events?  It would be help if you could catch the error when the offending call returns an error, and use that trigger to dump the error stack vi ERR_print_errors_fp or a similar call.  That would help illuminate what the client thinks the problem is.  There are about 50 places in the SSL stack where SSL_AD_UNEXPECTED_MESSAGE is asserted, and this would help us identify which case was being triggered

**Comment by mark99i** [comment]
@nhorman
I've encountered various errors. It all started with the Protocol Version error. I assumed that traffic dumps would help, but apparently not?
This happens at a random moment during the first few seconds of an active session.

Is there a way to redirect the openssl errors / stacktrace output to a separate file (as is done for SSLKEYLOGFILE, for example)?
I can compile openssl master (with latest python) and check, but I do not know how to deduce errors from it.

**Reply by nhorman** [reply]
You can use ERR_print_errors_fp() or ERR_print_errors() to write the errors to a bio or file pointer, but you're going to have to modify the code to do so, there is no environment variable to just print out all errors all the time.

If you don't want to modify the code you can run your application under gdb and place a breakpoint in ossl_rlayer_fatal.  When the error occurs you can print out the backtrace, and that will show us what condition is triggering the error.

**Reply by mark99i** [reply]
Thanks, I'll try. does it make sense to record traffic dumps and keys?

**Reply by nhorman** [reply]
yeah, it definitely helps, and will help us correlate the data being sent to the error being generated.  The problem as it exists currently is that the traffic dump shows us that the problem is happening, but not how the client is interpreting the data to trigger the alert response.  As I noted above, the unexpected message alert triggers in about 50 different locations (some of which are clearly irrelevant, like for dtls, but many are potentially sources of the error).  Theres no clear reason why the data in the dump would trigger this alert that I can see, so the client error log should definitively show us what code path is triggering the alert.  From there we can correlate the source of the error trigger to the data that is forcing said error.

**Comment by jsg49** [comment]
패킷 필터를 addr.ip 기준으로 변경하여 UDP 통신이 같이 되고 있는지 확인해 보시고, TCP로만 통신 되도록 설정을 변경해보시기 바랍니다.
